### PR TITLE
DSM Tests for each DSM Context Hashing Version

### PR DESCRIPTION
**Pull request by wconti27**
## Motivation

Adds the following tests:
- DSM Context Injection uses base64 encoded data
- DSM Context extraction can handle base64 encoded data

## Changes

New DSM Tests

## Workflow

1. ⚠️ Create your PR as draft ⚠️
2. Work on you PR until the CI passes (if something not related to your task is failing, you can ignore it)
3. Mark it as ready for review
    * Test logic is modified? -> Get a review from RFC owner. We're working on refining the `codeowners` file quickly.
    * Framework is modified, or non obvious usage of it -> get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

:rocket: Once your PR is reviewed, you can merge it!

🛟 [#apm-shared-testing](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X) 🛟

## Reviewer checklist

* [ ] [Relevant label](https://github.com/DataDog/system-tests/blob/main/docs/CI/labels.md) (`run-parametric-scenario`, `run-profiling-scenario`...) are presents
* [ ] No system-tests internal is modified. Otherwise, I have the approval from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)
* [ ] CI is green, or failing jobs are not related to this change (and you are 100% sure about this statement)
* [ ] A docker base image is modified?
    * [ ] the relevant `build-XXX-image` label is present
    * [ ] To R&P team: locally build and push the image to hub.docker.com 
* [ ] A scenario is added (or removed)?
    * [ ] Get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)
    * [ ] Once merged, add (or remove) it in system-test-dasboard nightly

### CODE REVIEW

File: tests/integrations/test_dsm.py [hunk 1]
```diff
@@ -370,3 +533,84 @@ def assert_checkpoint_presence(hash_, parent_hash, tags):
...
+            logging.info("System Tests RabbitMQ message using V1 DSM Pathway Encoding sent successfully")
+            return "ok"
+        except Exception as e:
+            logging.info(f"Error during DSM RabbitMQ publish message using V1 DSM Pathway Encoding: {e}")
+            return "error"
+
+    @staticmethod
+    def produce_rabbitmq_message_v2_base64_propagation(queue, exchange):
+        # Create a RabbitMQ client
+        conn = kombu.Connection("amqp://127.0.0.1:5672")
```
Comment (robertomonteromiguel) [inline tests/integrations/test_dsm.py:568]: I guess in this case it has no impact, but you should close the connection.

**Review by robertomonteromiguel — CHANGES_REQUESTED** [review]

File: tests/integrations/test_dsm.py [hunk 2]
```diff
@@ -328,6 +334,163 @@ def test_dsm_kinesis(self):
...
+        exchange = "dsm-propagation-test-injection-exchange"
+        edge_tags_out = ("direction:out", f"exchange:{exchange}", "has_routing_key:true", "type:rabbitmq")
+
+        message = self.consume_response["result"][0]
+
+        if "dd-pathway-ctx-base64" in message.headers:
+            encoded_pathway_b64 = message.headers["dd-pathway-ctx-base64"]
+
+            # assert that this is base64
+            assert base64.b64encode(base64.b64decode(encoded_pathway_b64)) == bytes(encoded_pathway_b64, "utf-8")
```
[the remark this answers is not in the document]
Comment (wconti27) [reply tests/integrations/test_dsm.py:389]: Yes, we that is fine.

File: tests/integrations/test_dsm.py (hunk 1 above)
Comment (wconti27) [reply tests/integrations/test_dsm.py:570]: updated to close all rabbitmq connections

**Review by robertomonteromiguel — APPROVED** [review]

File: tests/integrations/test_dsm.py [hunk 3]
```diff
@@ -328,6 +334,163 @@ def test_dsm_kinesis(self):
...
+                "consumer": 6884439977898629893,
+            },  # confirmed that this is correct consumer hash
+        }
+        producer_hash = language_hashes.get(context.library.library, language_hashes.get("default"))["producer"]
+        consumer_hash = language_hashes.get(context.library.library, language_hashes.get("default"))["consumer"]
+
+        queue = "dsm-propagation-test-v1-encoding-queue"
+        edge_tags_in = ("direction:in", f"topic:{queue}", "type:rabbitmq")
+
+        DsmHelper.assert_checkpoint_presence(
```
Comment (tabgok) [inline tests/integrations/test_dsm.py:455]: Are we checking the producer here?

File: tests/integrations/test_dsm.py [hunk 4]
```diff
@@ -370,3 +533,89 @@ def assert_checkpoint_presence(hash_, parent_hash, tags):
...
+            logging.info("System Tests RabbitMQ message using V1 DSM Pathway Encoding sent successfully")
+            conn.close()
+            return "ok"
+        except Exception as e:
+            logging.info(f"Error during DSM RabbitMQ publish message using V1 DSM Pathway Encoding: {e}")
+            conn.close()
+            return "error"
+
+    @staticmethod
+    def produce_rabbitmq_message_v2_base64_propagation(queue, exchange):
```
Comment (tabgok) [inline tests/integrations/test_dsm.py:568]: interesting, why are these written here instead of in the the weblog endpoint(s) ?

Comment (tabgok) [reply tests/integrations/test_dsm.py:568]: Is it because we don't have a mechanism to force the weblog to produce any given variant?

**Review by tabgok — APPROVED** [review]

File: tests/integrations/test_dsm.py (hunk 4 above)
Comment (wconti27) [reply tests/integrations/test_dsm.py]: (Discussed already privately but adding here): Yes, we can't force the weblog to produce a certain header. We also can't disable tracing, so the tracer will replace the added header anyways with a new context.

File: tests/integrations/test_dsm.py [hunk 5]
```diff
@@ -328,6 +334,163 @@ def test_dsm_kinesis(self):
...
+                hash_=producer_hash, parent_hash=0, tags=edge_tags_out,
+            )
+
+            # We autofail here since we anticipate the base64 header
+            assert 1 == 0, "Tracer should support DSM Base64 Pathway encoding"
+
+
+@features.datastreams_monitoring_support_for_v1_encoding
+@scenarios.integrations
+class Test_DsmContext_Extraction_V1:
```
[the remark this answers is not in the document]
Comment (wconti27) [reply tests/integrations/test_dsm.py]: Discussed already privately: We are doing two things here with these tests...

1. Testing that each tracer injects DSM context using base64 encoding when producing
2. Testing that each tracer is capable of reading base64 headers and ends up with the same context hash when consuming.

**Pull request merged**